### ISSUE TITLE
We may be able to reduce memory by using np.float16 instead of np.float32 matrices

### ISSUE THREAD

**Issue by charlesmartin14** [CalculatedContent/WeightWatcher#222]
When we detach the matrices from torch or TF and convert to numpy, I think the default converstion is to a float32 np matrix

This has been fixed in PyStateDictFileLayer using

```
  if type(weights)==torch.Tensor:
                    weights = weights.data.cpu().detach().float().numpy().astype(np.float16)
                    if biases is not None:
                        biases = biases.data.cpu().detach().float().numpy().astype(np.float16)
```

We probably want this everywhere

**Comment by charlesmartin14** [comment]
This has been implemented as


```
"""This method is over-written in tourch cuda is available"""
torch_T_to_np = lambda T: T.to("cpu").float().numpy()

# 
try:
    import torch
    if torch.cuda.is_available() :
        logger.info("Torch CUDA available, using torch_wrapper")

        # float changed to half to save memory
        torch_T_to_np = lambda T: T.to("cpu").half().numpy()
...
```

Needs more testing
